libsupport: Improve compatibility with nvcc

ddn0 opened the pull request:
nvcc (versions 11.1, 11.3 at least) does not allow non-trivial classes
to be defined inside statement expressions, which leads to errors like:

  a class that is not trivially copyable cannot be defined inside a statement expression

when using KATANA_CHECKED in any file that nvcc could see. Besides
actual CUDA files, this category includes any header files included by
CUDA files. nvcc also has trouble recognizing implicit conversions to
katana::Result.

Workaround these problems point-wise by avoiding KATANA_CHECKED and
implicit conversions to katana::Result.

ddn0 commented on lonestar/analytics/distributed/bfs/CMakeLists.txt:
```diff
@@ -3,7 +3,6 @@ app_dist(bfs_push bfs-push)
-#add_test_dist(bfs-pull-dist INPUT rmat15 INPUT_URI ${BASEINPUT}/scalefree/rmat15.gr -graphTranspose=${BASEINPUT}/scalefree/transpose/rmat15.tgr)
-#add_test_dist(bfs-pull-dist INPUT ldbc_003 INPUT_URI ${BASEINPUT}/propertygraphs/ldbc_003/meta --edgePropertyName="classYear" -startNode=134)
+add_test_dist(bfs-pull-dist INPUT rmat15 INPUT_URI ${BASEINPUT}/propertygraphs/rmat15/meta --edgePropertyName=value)
+add_test_dist(bfs-pull-dist STORE_PART INPUT rmat15 INPUT_URI ${BASEINPUT}/propertygraphs/rmat15/meta --edgePropertyName=value)
```
your editor seems to like removing newlines at the end of a file.

I'd figure out some way to at least preserve a final newline if present. Or just always add a newline (which is common for most POSIX text editors). Otherwise, you will end up with a bunch of spurious changes in your commits.

gurbinder533 replied:
Fixed.

tylershunt commented on libcusp/include/galois/graphs/Distribution.h:
```diff
@@ -252,6 +253,117 @@ class Distribution {
+   * GraphTranpose does in-memory transpose for pull-style
+   * algorithms to iterate over in-coming edges. This function
+   * is called if opt_.transpose is true.
+   */
```
I think this could also be called if `opt_.transpose` is false and the graph in storage happened to be transposed right?

roshandathathri replied:
We are not planning to store transposed files in storage.

gurbinder533 replied:
The current state is that whatever topology was used when graph is partitioned for the first time, gets stored on disk. And then based on the next use case, the topology is transposed or not.

tylershunt reviewed (commented):
Changes to `DistGraph.h` look good; that the only bit I really understand though.

roshandathathri commented on libcusp/include/galois/graphs/Distribution.h:
```diff
@@ -252,6 +253,117 @@ class Distribution {
+   * algorithms to iterate over in-coming edges. This function
+   * is called if opt_.transpose is true.
+   */
+  Result<void> GraphTranspose() {
```
This is a strange function for this class. Why can't this be in DistGraph?

roshandathathri commented on libgalois/src/PropertyFileGraph.cpp:
```diff
@@ -62,16 +62,15 @@ MapTopology(const tsuba::FileView& file_view) {
   }
 
-  const uint64_t* out_indices = &data[4];
+  uint64_t* out_indices = const_cast<uint64_t*>(&data[4]);
```
Seems innocuous.

gurbinder533 replied:
We need the mutableBuffers for arrow arrays.

roshandathathri commented on lonestar/analytics/distributed/pagerank/CMakeLists.txt:
```diff
@@ -1,5 +1,6 @@
 app_dist(pagerank_pull pagerank-pull)
 #add_test_dist(pagerank-pull-dist INPUT rmat15 INPUT_URI ${BASEINPUT}/scalefree/rmat15.gr -graphTranspose=${BASEINPUT}/scalefree/transpose/rmat15.tgr -maxIterations=100)
+add_test_dist(pagerank-pull-dist INPUT rmat15 INPUT_URI ${BASEINPUT}/propertygraphs/rmat15 --edgePropertyName=value -maxIterations=100)
```
Drop above line

roshandathathri reviewed (approved):
LGTM

Why not enable pull-style variants for all applications?

tylershunt reviewed (approved):
LGTM

roshandathathri commented:
Related to #503: if the partitions are saved after executing pull-style algorithms, then transpose graph will be saved. Does storing the RDG maintain whether the topology is in CSR or in CSC?

gurbinder533 commented:
> Related to #503: if the partitions are saved after executing pull-style algorithms, then transpose graph will be saved. Does storing the RDG maintain whether the topology is in CSR or in CSC?

It does. The transposed flag is stored in the RDG and that is how we know if we need to transpose it or not.

ddn0 committed:
libsupport: Improve compatibility with nvcc

nvcc (versions 11.1, 11.3 at least) does not allow non-trivial classes
to be defined inside statement expressions, which leads to errors like:

  a class that is not trivially copyable cannot be defined inside a statement expression

when using KATANA_CHECKED in any file that nvcc could see. Besides
actual CUDA files, this category includes any header files included by
CUDA files. nvcc also has trouble recognizing implicit conversions to
katana::Result.

Workaround these problems point-wise by avoiding KATANA_CHECKED and
implicit conversions to katana::Result.

witchel commented on libsupport/include/katana/EntityTypeManager.h:
```diff
@@ -122,8 +122,12 @@ class KATANA_EXPORT EntityTypeManager {
 
-    TypeProperties type_properties = KATANA_CHECKED(
-        DoAssignEntityTypeIDsFromProperties(properties, entity_type_manager));
+    auto res =
+        DoAssignEntityTypeIDsFromProperties(properties, entity_type_manager);
```
Thanks Donald!  You might want to add a comment here; otherwise, an enterprising hacker might replace this construct with a `KATANA_CHECKED `macro...

ddn0 replied:
Sure! I can add that. Do you mind approving this if it looks good to you? This is blocking some other work.

witchel replied:
Sure thing.

scober commented:
Thank you!
